File: filly-core/src/widgets/multiselect.rs
use std::collections::HashSet;
use crate::event::{Event, KeyCode};
use crate::render::{
    Alignment, BorderStyle, EdgeInsets, ListItem, Rect, RenderNode, RenderTree, TextStyle,
};
use crate::widget::{EventResult, Widget};
use filly_protocol::WidgetResponse;

pub struct MultiselectWidget {
    pub title: String,
    pub message: String,
    pub choices: Vec<String>,
    pub selected: HashSet<usize>,
    pub cursor: usize,
    pub query: String,
    pub placeholder: String,
    pub min: usize,
    pub max: usize,
    dirty: bool,
}

impl MultiselectWidget {
    pub fn new(
        title: String,
        message: String,
        choices: Vec<String>,
        placeholder: Option<String>,
        min: Option<usize>,
        max: Option<usize>,
    ) -> Self {
        Self {
            title,
            message,
            choices,
            selected: HashSet::new(),
            cursor: 0,
            query: String::new(),
            placeholder: placeholder.unwrap_or("Type to filter...".into()),
            min: min.unwrap_or(0),
            max: max.unwrap_or(usize::MAX),
            dirty: true,
        }
    }

    fn filtered(&self) -> Vec<usize> {
        if self.query.is_empty() {
            (0..self.choices.len()).collect()
        } else {
            let q = self.query.to_lowercase();
            self.choices
                .iter()
                .enumerate()
                .filter(|(_, c)| c.to_lowercase().contains(&q))
                .map(|(i, _)| i)
                .collect()
        }
    }
}

impl Widget for MultiselectWidget {
    fn render(&self, area: Rect) -> RenderTree {
        let box_w = ((area.width as f32 * 0.7) as u16).min(area.width.saturating_sub(2));
        let box_h = ((area.height as f32 * 0.8) as u16).min(area.height.saturating_sub(2));
        let box_x = (area.width.saturating_sub(box_w)) / 2;
        let box_y = (area.height.saturating_sub(box_h)) / 2;

        let inner_h = box_h.saturating_sub(2);
        let filtered = self.filtered();
        let absolute_selected = filtered.get(self.cursor).copied();
        let mut children = Vec::new();

        let mut y = 1;

        if !self.title.is_empty() {
            children.push(RenderNode::Text {
                rect: Rect::new(1, y, box_w - 2, 1),
                content: self.title.clone(),
                alignment: Alignment::Center,
                role: None, label: None, description: None, style: TextStyle { bold: true, fg: Some("green".into()), ..TextStyle::normal() },
            });
            y += 1;
        }
        if !self.message.is_empty() {
            children.push(RenderNode::Text {
                rect: Rect::new(1, y, box_w - 2, 2),
                content: self.message.clone(),
                alignment: Alignment::Left,
                style: TextStyle::normal(), role: None, label: None, description: None,
            });
            y += 2;
        }

        children.push(RenderNode::Input {
            rect: Rect::new(1, y, box_w - 2, 1),
            text: if self.query.is_empty() { self.placeholder.clone() } else { format!("> {}", self.query) },
            cursor: if self.query.is_empty() { 0 } else { 2 + self.query.len() },
            placeholder: self.placeholder.clone(),
            masked: false, role: None, label: None, description: None,
        });
        y += 1;

        let list_h = inner_h.saturating_sub(y).saturating_sub(2);

        let items: Vec<ListItem> = self.choices.iter().enumerate().map(|(i, c)| {
            let mark = if self.selected.contains(&i) { "[x]" } else { "[ ]" };
            ListItem { label: format!(" {} {}", mark, c), meta: None }
        }).collect();

        children.push(RenderNode::List {
            rect: Rect::new(1, y, box_w - 2, list_h),
            items,
            selected: absolute_selected,
            highlight_style: TextStyle::selected(), role: None, label: None, description: None,
        });

        children.push(RenderNode::Text {
            rect: Rect::new(1, inner_h, box_w - 2, 1),
            content: format!(" Selected: {}/{}   Space=toggle  Enter=confirm  Esc=cancel",
                self.selected.len(), self.choices.len()),
            alignment: Alignment::Left,
            style: TextStyle::muted(), role: None, label: None, description: None,
        });
        children.push(RenderNode::Text {
            rect: Rect::new(1, inner_h + 1, box_w - 2, 1),
            content: "Type:filter  Up/Down:move  Space:toggle  Enter:confirm  Esc:cancel".into(),
            alignment: Alignment::Center,
            style: TextStyle::muted(), role: None, label: None, description: None,
        });

        RenderTree::Container {
            rect: Rect::new(box_x, box_y, box_w, box_h),
            background: None,
            border: BorderStyle::Single,
            padding: EdgeInsets::zero(), role: None, label: None, description: None,
            children,
        }
    }

    fn handle_event(&mut self, event: &Event) -> EventResult {
        match event {
            Event::Key(code, _) => match code {
                KeyCode::Esc => EventResult::Response(WidgetResponse {
                    result: None,
                    cancelled: true,
                    error: None,
                }),
                KeyCode::Up => {
                    if self.cursor > 0 {
                        self.cursor -= 1;
                        self.dirty = true;
                    }
                    EventResult::Handled
                }
                KeyCode::Down => {
                    let filtered = self.filtered();
                    if self.cursor + 1 < filtered.len() {
                        self.cursor += 1;
                        self.dirty = true;
                    }
                    EventResult::Handled
                }
                KeyCode::Char(' ') => {
                    let filtered = self.filtered();
                    if let Some(&idx) = filtered.get(self.cursor) {
                        if self.selected.contains(&idx) {
                            if self.selected.len() > self.min {
                                self.selected.remove(&idx);
                                self.dirty = true;
                            }
                        } else if self.selected.len() < self.max {
                            self.selected.insert(idx);
                            self.dirty = true;
                        }
                    }
                    EventResult::Handled
                }
                KeyCode::Enter => {
                    if self.selected.len() >= self.min && self.selected.len() <= self.max {
                        let selected: Vec<String> = self
                            .selected
                            .iter()
                            .map(|&i| self.choices[i].clone())
                            .collect();
                        EventResult::Response(WidgetResponse {
                            result: Some(serde_json::Value::Array(
                                selected
                                    .into_iter()
                                    .map(serde_json::Value::String)
                                    .collect(),
                            )),
                            cancelled: false,
                            error: None,
                        })
                    } else {
                        EventResult::Handled
                    }
                }
                KeyCode::Char(c) => {
                    self.query.push(*c);
                    self.cursor = 0;
                    self.dirty = true;
                    EventResult::Handled
                }
                KeyCode::Backspace => {
                    self.query.pop();
                    self.cursor = 0;
                    self.dirty = true;
                    EventResult::Handled
                }
                _ => EventResult::Unhandled,
            },
            _ => EventResult::Unhandled,
        }
    }

    fn is_dirty(&self) -> bool {
        self.dirty
    }
    fn clear_dirty(&mut self) {
        self.dirty = false;
    }
}

File: filly-core/widgets/multiselect.c
#include <stdio.h>
#include "multiselect.h"
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

typedef struct {
    char *title;
    char *message;
    char **choices;
    int count;
    char *query;
    int cursor;
    bool *selected;
    char *placeholder;
    int min_sel;
    int max_sel;
    int *filtered;
    int filtered_count;
    bool dirty;
} MultiselectData;

static void ms_update(MultiselectData *d) {
    free(d->filtered);
    if (strlen(d->query) == 0) {
        d->filtered = malloc(d->count * sizeof(int));
        d->filtered_count = d->count;
        for (int i = 0; i < d->count; i++) d->filtered[i] = i;
    } else {
        d->filtered = malloc(d->count * sizeof(int));
        d->filtered_count = 0;
        char *ql = strdup(d->query);
        for (int i = 0; ql[i]; i++) ql[i] = tolower(ql[i]);
        for (int i = 0; i < d->count; i++) {
            char *cl = strdup(d->choices[i]);
            for (int j = 0; cl[j]; j++) cl[j] = tolower(cl[j]);
            if (strstr(cl, ql)) d->filtered[d->filtered_count++] = i;
            free(cl);
        }
        free(ql);
    }
    if (d->cursor >= d->filtered_count) d->cursor = d->filtered_count > 0 ? d->filtered_count - 1 : 0;
}

static void multiselect_render(Widget *self, Rect area, RenderTree *out) {
    MultiselectData *d = (MultiselectData *)(self + 1);
    ms_update(d);
    memset(out, 0, sizeof(*out));
    int box_w = (int)(area.w * 0.7f);
    if (box_w > area.w - 2) box_w = area.w - 2;
    int box_h = (int)(area.h * 0.8f);
    if (box_h > area.h - 2) box_h = area.h - 2;
    int box_x = (area.w - box_w) / 2, box_y = (area.h - box_h) / 2;

    out->accessible.role = strdup("dialog");
    out->accessible.label = strdup(d->title ? d->title : "Multiselect");

    RenderTree *children = calloc(5 + 1, sizeof(RenderTree));
    int idx = 0;
    if (d->title && strlen(d->title)) {
        children[idx].type = RNODE_TEXT;
        children[idx].rect = rect_new(1, 0, box_w - 2, 1);
        children[idx].text.content = strdup(d->title);
        children[idx].text.align = ALIGN_CENTER;
        children[idx].text.style = textstyle_selected();
        children[idx].accessible.role = strdup("heading");
        children[idx].accessible.label = strdup(d->title);
        idx++;
    }
    int y = 1;
    if (d->message && strlen(d->message)) {
        children[idx].type = RNODE_TEXT;
        children[idx].rect = rect_new(1, y, box_w - 2, 2);
        children[idx].text.content = strdup(d->message);
        children[idx].text.align = ALIGN_LEFT;
        children[idx].text.style = textstyle_normal();
        children[idx].accessible.role = strdup("description");
        idx++; y += 2;
    }
    children[idx].type = RNODE_INPUT;
    children[idx].rect = rect_new(1, y, box_w - 2, 1);
    children[idx].input.text = strlen(d->query) ? malloc(strlen(d->query) + 3) : strdup("");
    if (strlen(d->query)) sprintf(children[idx].input.text, "> %s", d->query);
    children[idx].input.cursor = strlen(d->query) + 2;
    children[idx].input.placeholder = d->placeholder ? strdup(d->placeholder) : strdup("Type to filter...");
    children[idx].input.masked = false;
    children[idx].accessible.role = strdup("searchbox");
    children[idx].accessible.label = strdup(d->placeholder ? d->placeholder : "Filter");
    idx++; y++;

    int list_h = box_h - y - 2;
    children[idx].type = RNODE_LIST;
    children[idx].rect = rect_new(1, y, box_w - 2, list_h > 0 ? list_h : 1);
    children[idx].list.item_count = d->filtered_count;
    children[idx].list.items = malloc(d->filtered_count * sizeof(ListItem));
    for (int i = 0; i < d->filtered_count; i++) {
        char label[1024];
        int orig = d->filtered[i];
        snprintf(label, sizeof(label), " %s %s", d->selected[orig] ? "[x]" : "[ ]", d->choices[orig]);
        children[idx].list.items[i] = listitem_new(label);
    }
    children[idx].list.selected = d->cursor;
    children[idx].list.highlight = textstyle_selected();
    children[idx].accessible.role = strdup("listbox");
    idx++;

    int sel_count = 0;
    for (int i = 0; i < d->count; i++) if (d->selected[i]) sel_count++;
    children[idx].type = RNODE_TEXT;
    children[idx].rect = rect_new(1, box_h - 2, box_w - 2, 1);
    char footer_text[256];
    snprintf(footer_text, sizeof(footer_text), "Selected: %d/%d  Space:toggle  Enter:confirm  Esc:cancel", sel_count, d->count);
    children[idx].text.content = strdup(footer_text);
    children[idx].text.align = ALIGN_CENTER;
    children[idx].text.style = textstyle_muted();
    idx++;

    out->type = RNODE_CONTAINER;
    out->rect = rect_new(box_x, box_y, box_w, box_h);
    out->container.border = BORDER_SINGLE;
    out->container.padding = edgeinsets_zero();
    out->container.children = children;
    out->container.child_count = idx;
}

static EventResult multiselect_handle_event(Widget *self, Event *ev, Backend *backend) {
    (void)backend;
    MultiselectData *d = (MultiselectData *)(self + 1);
    if (ev->type != EVENT_KEY) return event_result_unhandled();
    switch (ev->code) {
        case KEY_ESC:
            return event_result_response((WidgetResponse){ .result = NULL, .cancelled = true, .error = NULL });
        case KEY_UP:
            if (d->cursor > 0) { d->cursor--; d->dirty = true; }
            return event_result_handled();
        case KEY_DOWN:
            if (d->cursor + 1 < d->filtered_count) { d->cursor++; d->dirty = true; }
            return event_result_handled();
        case KEY_CHAR:
            if (ev->ch == ' ') {
                if (d->filtered_count > 0 && d->cursor < d->filtered_count) {
                    int orig = d->filtered[d->cursor];
                    int sel_count = 0;
                    for (int i = 0; i < d->count; i++) if (d->selected[i]) sel_count++;
                    if (d->selected[orig]) { if (sel_count > d->min_sel) d->selected[orig] = false; }
                    else { if (sel_count < d->max_sel) d->selected[orig] = true; }
                    d->dirty = true;
                }
            } else {
                int len = strlen(d->query);
                d->query = realloc(d->query, len + 2);
                d->query[len] = ev->ch; d->query[len + 1] = '\0';
                d->cursor = 0; d->dirty = true;
            }
            return event_result_handled();
        case KEY_BACKSPACE:
            if (strlen(d->query) > 0) { d->query[strlen(d->query) - 1] = '\0'; d->cursor = 0; d->dirty = true; }
            return event_result_handled();
        case KEY_ENTER: {
            int sel_count = 0;
            for (int i = 0; i < d->count; i++) if (d->selected[i]) sel_count++;
            if (sel_count >= d->min_sel && sel_count <= d->max_sel) {
                cJSON *arr = cJSON_CreateArray();
                for (int i = 0; i < d->count; i++)
                    if (d->selected[i]) cJSON_AddItemToArray(arr, cJSON_CreateString(d->choices[i]));
                return event_result_response((WidgetResponse){ .result = arr, .cancelled = false, .error = NULL });
            }
            return event_result_handled();
        }
        default: return event_result_unhandled();
    }
}

static bool multiselect_is_dirty(Widget *self) { return ((MultiselectData *)(self + 1))->dirty; }
static void multiselect_clear_dirty(Widget *self) { ((MultiselectData *)(self + 1))->dirty = false; }
static void multiselect_destroy(Widget *self) {
    MultiselectData *d = (MultiselectData *)(self + 1);
    free(d->title); free(d->message); free(d->query); free(d->placeholder);
    for (int i = 0; i < d->count; i++) free(d->choices[i]);
    free(d->choices); free(d->selected); free(d->filtered);
}

Widget *multiselect_widget_new(const char *title, const char *message, char **choices, int count,
                                const char *placeholder, int min_sel, int max_sel) {
    Widget *w = calloc(1, sizeof(Widget) + sizeof(MultiselectData));
    w->vtable.render = multiselect_render;
    w->vtable.handle_event = multiselect_handle_event;
    w->vtable.is_dirty = multiselect_is_dirty;
    w->vtable.clear_dirty = multiselect_clear_dirty;
    w->vtable.destroy = multiselect_destroy;
    MultiselectData *d = (MultiselectData *)(w + 1);
    d->title = title ? strdup(title) : NULL;
    d->message = message ? strdup(message) : NULL;
    d->choices = malloc(count * sizeof(char *));
    for (int i = 0; i < count; i++) d->choices[i] = strdup(choices[i]);
    d->count = count;
    d->query = strdup("");
    d->cursor = 0;
    d->selected = calloc(count, sizeof(bool));
    d->placeholder = placeholder ? strdup(placeholder) : strdup("Type to filter...");
    d->min_sel = min_sel;
    d->max_sel = max_sel > 0 ? max_sel : count;
    d->filtered = NULL;
    d->filtered_count = 0;
    d->dirty = true;
    return w;
}